drivers/java/src/main/java/com/rethinkdb/net/SocketWrapper.java

```diff
@@ -2,47 +2,125 @@
 
 public class SocketWrapper {
-    private SocketChannel socketChannel;
-    private Optional<Long> timeout = Optional.empty();
+    /**
+     * Default SSL/TLS protocol version
+     * <p>
+     * This property is defined as String {@value #DEFAULT_SSL_PROTOCOL}
+     */
+    private static final String DEFAULT_SSL_PROTOCOL = "TLSv1.2";
+
+    final ReentrantLock mu = new ReentrantLock();
+
+    private Socket socket = null;
+
+    private SocketFactory socketFactory = SocketFactory.getDefault();
+    private SSLSocket sslSocket = null;
 
-    public final String hostname;
-    public final int port;
+    protected OutputStream writeStream = null;
+    protected InputStream readStream = null;
 
-    public SocketWrapper(String hostname, int port,
+    // options
+    private Optional<InputStream> certFile = Optional.empty();
+    private Optional<SSLContext> sslContext = Optional.empty();
+    private Optional<Long> timeout = Optional.empty();
+    private final String hostname;
+    private final int port;
+
+    public SocketWrapper(String hostname,
+                         int port,
+                         Optional<InputStream> certFile,
+                         Optional<SSLContext> sslContext,
                          Optional<Long> timeout) {
         this.hostname = hostname;
         this.port = port;
+        this.certFile = certFile;
+        this.sslContext = sslContext;
         this.timeout = timeout;
-        try {
-            this.socketChannel = SocketChannel.open();
-        } catch (IOException e) {
-            throw new ReqlDriverError(e);
-        }
     }
 
     public void connect(ByteBuffer handshake) throws TimeoutException {
         Optional<Long> deadline = timeout.map(Util::deadline);
+
+        mu.lock();
         try {
-            socketChannel.configureBlocking(true);
-            socketChannel.setOption(StandardSocketOptions.TCP_NODELAY, true);
-            socketChannel.setOption(StandardSocketOptions.SO_KEEPALIVE, true);
-            socketChannel.socket().connect(
-                    new InetSocketAddress(hostname, port),
-                    timeout.orElse(0L).intValue());
-            socketChannel.write(handshake);
-            String msg = readNullTerminatedString(deadline);
+            // establish connection
+            final InetSocketAddress addr = new InetSocketAddress(hostname, port);
+            socket = socketFactory.createSocket();
+            socket.connect(addr, timeout.orElse(0L).intValue());
+            socket.setTcpNoDelay(true);
+            socket.setKeepAlive(true);
+
+            // should we secure the connection?
+            if (certFile.isPresent() || sslContext.isPresent()) {
+                // is certFile provided? if so, it has precedence over SSLContext
+                if (certFile.isPresent()) {
+                    try {
+                        final CertificateFactory cf = CertificateFactory.getInstance("X.509");
```

Comment:
Could you move this part into the builder, so just the SSLContext is passed to the connection constructor?